### PR TITLE
[Feature] Extract wide common factors

### LINKED ISSUE #6082: [Feature] Extract wide common factors
**Is your feature request related to a problem? Please describe.**

If the where expression contains consecutive or clauses, the predicate can only be executed in higher-level operators. such as
```
select * from a, b where (1<a.k1<3 and b.k1 in ('a')) or (2<a.k1<4 and b.k1 in ('b'))
```
Because the two or clauses of where involve two tables respectively. So the expression can only be calculated in the join operator. This will cause the query to scan all tables a and b.

In fact, although these two clauses do not have a common factor, we can extract a wider range to filter the data in tables a and b in advance. Reduce the number of filtered rows.
Finally, perform precise filtering on the join operator.
such as:
For a.k1, the ranges in the two clauses can be combined to obtain:
1<a.k1<4, in this way, table a can be filtered in advance to reduce the amount of join data in table a b.

After rewriting, the query looks like:
```
select * from a,b where (1<a.k1<4) and (b.k1 in('a','b')) and (1<a.k1<3 and b.k1 in ('a' )) or (2<a.k1<4 and b.k1 in ('b'))
```

### PR DESCRIPTION
**Pull request by EmmyMiao87**
## Proposed changes
This PR mainly adds a rewrite rule 'ExtractCommonFactorsRule'
  used to extract wide common factors in the planning stage for 'Expr'.
The main purpose of this rule is to extract (Range or In) expressions
  that can be combined from each or clause.
E.g:
  Origin expr: (1<a<3 and b in ('a') ) or (2<a<4 and b in ('b'))
  Rewritten expr: (1<a<4 ) and (b in ('a', 'b')) and ((1<a<3 and b in ('a') ) or (2<a<4 and b in ('b')))
Although the range of the wide common factors is larger than the real range,
  the wide common factors only involve a single column, so it can be pushed down to the scan node,
  thereby reducing the amount of scanned data in advance and improving the query speed.

It should be noted that this optimization strategy is not for all scenarios.
When filter rate of the wide common factor is too low,
  the query will consume an extra time to calculate the wide common factors.

So this strategy can be switched by configuring session variables 'extract_wide_range_expr'.
The default policy is enabled which means this strategy takes effect.
If you encounter unsatisfactory filtering rate, you can set the variable to false.
It will turn off the strategy.

Fixed #6082

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] New feature (non-breaking change which adds functionality)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #6082) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

### CODE REVIEW

File: docs/zh-CN/administrator-guide/variables.md [hunk 1]
```diff
@@ -383,3 +383,7 @@ SELECT /*+ SET_VAR(query_timeout = 1, enable_partition_cache=true) */ sleep(3);
     在一个排序的查询之中，一个上层节点接收下层节点有序数据时，会在exchange node上进行对应的排序来保证最终的数据是有序的。但是单线程进行多路数据归并时，如果数据量过大，会导致exchange node的单点的归并瓶颈。
 
     Doris在这部分进行了优化处理，如果下层的数据节点过多。exchange node会启动多线程进行并行归并来加速排序过程。该参数默认为False，即表示 exchange node 不采取并行的归并排序，来减少额外的CPU和内存消耗。
+
+* `extract_wide_range_expr`
```
Comment (yangzhg) [inline docs/zh-CN/administrator-guide/variables.md:387]: `extract_wide_range_expr` is the name of who this implementation, not the name of the function， you may change the name

Comment (EmmyMiao87) [reply docs/zh-CN/administrator-guide/variables.md:387]: Maybe extract_common_factors?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java [hunk 2]
```diff
@@ -36,11 +37,17 @@
 public class ExprRewriter {
     private int numChanges_ = 0;
     private final List<ExprRewriteRule> rules_;
+    private List<ExprRewriteRule> onceRules_ = Lists.newArrayList();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:40]: Add comment to explain what is `onceRules`

**Review by yangzhg — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java (hunk 2 above)
Comment (EmmyMiao87) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/ExprRewriter.java:49]: OK

**Pull request merged**